File: app/src/main/java/net/squanchy/navigation/deeplink/RoutingException.java
package net.squanchy.navigation.deeplink;

public class RoutingException extends RuntimeException {

    static RoutingException malformedUrl(String url, Throwable cause) {
        return new RoutingException("Malformed URL: \"" + url + "\"", cause);
    }

    static RoutingException unhandledScheme(String url) {
        return new RoutingException("The URL \"" + url + "\" doesn't have a valid scheme");
    }

    static RoutingException unhandledPath(String url) {
        return new RoutingException("The URL \"" + url + "\" doesn't match any valid link target");
    }

    private RoutingException(String message) {
        super(message);
    }

    private RoutingException(String message, Throwable cause) {
        super(message, cause);
    }
}

File: app/src/main/java/net/squanchy/navigation/deeplink/RoutingException.kt
package net.squanchy.navigation.deeplink

internal class RoutingException : RuntimeException {

    private constructor(message: String) : super(message)

    private constructor(message: String, cause: Throwable) : super(message, cause)

    companion object {

        fun malformedUrl(url: String, cause: Throwable): RoutingException {
            return RoutingException("Malformed URL: \"$url\"", cause)
        }

        fun unhandledScheme(url: String): RoutingException {
            return RoutingException("The URL \"$url\" doesn't have a valid scheme")
        }

        fun unhandledPath(url: String): RoutingException {
            return RoutingException("The URL \"$url\" doesn't match any valid link target")
        }
    }
}
